### PR TITLE
first attempt at fuzz testing

### PR DESCRIPTION
**Pull request by lastmjs**
-   [x] Breaking changes enumerated in [the release issue](https://github.com/demergent-labs/azle/issues/2053)
-   [x] New documentation enumerated in [the release issue](https://github.com/demergent-labs/azle/issues/2053)
-   [x] Code has been declaratized
-   [x] Error handling beautiful (no unwraps or expects etc)
-   [x] Code tested thoroughly

### CODE REVIEW

File: src/build/stable/commands/compile/wasm_binary/rust/stable_canister_template/src/ic/call_raw.rs [hunk 1]
```diff
@@ -33,14 +48,31 @@ pub fn get_function(ctx: Ctx) -> QuickJsResult<Function> {
...
+                // thus allowing us to recover from a trap and persist that state
+                let _cleanup = scopeguard::guard((), |_| {
+                    let reject_id = format!("_reject_{}", &promise_id);
+                    let resolve_id = format!("_resolve_{}", &promise_id);
+
+                    let globals = ctx.clone().globals();
 
-                let result = quickjs_with_ctx(|ctx| {
-                    resolve_or_reject(ctx.clone(), &call_result, &promise_id)?;
+                    let reject_ids = globals.get::<_, Object>("_azleRejectIds").unwrap();
```
Comment (bdemann) [inline src/build/stable/commands/compile/wasm_binary/rust/stable_canister_template/src/ic/call_raw.rs:65]: Can you fix these four unwraps?

File: test/index.ts [hunk 2]
```diff
@@ -146,20 +238,35 @@ function processEnvVars(): {
     shouldRunTests: boolean;
     shouldRunTypeChecks: boolean;
     shouldRecordBenchmarks: boolean;
+    shouldFuzz: boolean;
 } {
     const runTests = process.env.AZLE_RUN_TESTS ?? 'true';
     const runTypeChecks = process.env.AZLE_RUN_TYPE_CHECKS ?? 'true';
     const recordBenchmarks = process.env.AZLE_RECORD_BENCHMARKS ?? 'false';
+    const fuzz = process.env.AZLE_FUZZ ?? 'false';
```
Comment (bdemann) [inline test/index.ts:246]: Can we rename this to runFuzzTests and AZLE_RUN_FUZZ_TESTS?

File: test/index.ts [hunk 3]
```diff
@@ -57,6 +63,92 @@ export function runTests(
                 runBenchmarksForCanisters(canisterNamesArray));
         });
     }
+
+    // TODO we should work on what the UI looks like if there are multiple of these things running
```
Comment (bdemann) [inline test/index.ts:67]: Get rid of this old TODO

File: test/index.ts [hunk 4]
```diff
@@ -57,6 +63,92 @@ export function runTests(
...
+    if (shouldFuzz && canisterNames !== undefined) {
+        // TODO right now we just test the first canister...that might be good enough for now?
```
Comment (bdemann) [inline test/index.ts:69]: Are these TODOs related to and outdated with the one above?

File: test/index.ts [hunk 5]
```diff
@@ -57,6 +63,92 @@ export function runTests(
...
+        // TODO because I feel that the first canister will generally call the other canisters
+        describe(`fuzz`, () => {
+            it('runs fuzz tests for all canisters', () => {
+                // execSyncPretty(`cuzz`, 'inherit');
+                const dfxFile = readFileSync(
+                    join(process.cwd(), 'dfx.json'),
+                    'utf-8'
+                );
+                console.log(dfxFile);
```
Comment (bdemann) [inline test/index.ts:78]: info?

File: src/build/stable/commands/compile/wasm_binary/rust/stable_canister_template/src/ic/call_raw.rs (hunk 1 above)
Comment (lastmjs) [reply src/build/stable/commands/compile/wasm_binary/rust/stable_canister_template/src/ic/call_raw.rs]: Done

File: test/index.ts (hunk 2 above)
Comment (lastmjs) [reply test/index.ts:158]: I like it how it is, we discussed this a bit, we can discuss more.

File: test/index.ts (hunk 3 above)
Comment (lastmjs) [reply test/index.ts]: Done

File: test/index.ts (hunk 4 above)
Comment (lastmjs) [reply test/index.ts]: All removed and refactored

File: test/index.ts (hunk 5 above)
Comment (lastmjs) [reply test/index.ts]: Gone

File: test/index.ts [hunk 6]
```diff
@@ -57,6 +63,92 @@ export function runTests(
...
         });
     }
+
+    // TODO we should work on what the UI looks like if there are multiple of these things running
+    if (shouldFuzz && canisterNames !== undefined) {
+        // TODO right now we just test the first canister...that might be good enough for now?
+        // TODO because I feel that the first canister will generally call the other canisters
+        describe(`fuzz`, () => {
+            it('runs fuzz tests for all canisters', () => {
+                // execSyncPretty(`cuzz`, 'inherit');
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply test/index.ts]: Gone

File: src/build/experimental/commands/compile/wasm_binary/rust/experimental_canister_template/src/lib.rs [hunk 7]
```diff
@@ -95,8 +90,3 @@ pub fn _azle_get_benchmarks() -> Vec<benchmarking::BenchmarkEntry> {
     benchmarking::BENCHMARKS_REF_CELL
         .with(|benchmarks_ref_cell| benchmarks_ref_cell.borrow().clone())
 }
-
-#[ic_cdk_macros::query]
```
Comment (bdemann) [inline src/build/experimental/commands/compile/wasm_binary/rust/experimental_canister_template/src/lib.rs]: What this function only around to fix the memory leak? If we have another memory leak we will add it, or something similar back? Just curious

File: tests/end_to_end/candid_rpc/functional_syntax/motoko_examples/http_counter/src/index.ts [hunk 8]
```diff
@@ -185,7 +185,7 @@ export default Canister({
                     };
                 }
                 default: {
-                    return ic.trap('unreachable');
+                    return ic.trap('should not happen');
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/functional_syntax/motoko_examples/http_counter/src/index.ts:188]: you never did like the word unreachable did you? Is that all that's going on here?

File: test/fuzz.ts
Comment (bdemann) [inline test/fuzz.ts:1]: nice and declarative

File: test/index.ts (hunk 2 above)
Comment (bdemann) [reply test/index.ts:158]: I'll leave further discussion for when we tackle https://github.com/demergent-labs/azle/issues/2286

**Review by bdemann — APPROVED** [review]
